components/string.rst

```diff
@@ -0,0 +1,378 @@
+.. index::
+   single: String
+   single: Components; String
+
+The String Component
+====================
+
+    The String component provides a single object-oriented API to work with
+    three "unit systems" of strings: bytes, code points and grapheme clusters.
+
+Installation
+------------
+
+.. code-block:: terminal
+
+    $ composer require symfony/string
+
+.. include:: /components/require_autoload.rst.inc
+
+What is a String?
+-----------------
+
+You can skip this section if you already know what a *"code point"* or a
+*"grapheme cluster"* are in the context of handling strings. Otherwise, read
+this section to learn about the terminology used by this component.
+
+Languages like English require a very limited set of characters and symbols to
+display any content. Each string is a series of characters (letters or symbols)
+and they can be encoded even with the most limited standards (e.g. `ASCII`_).
+
+However, other languages require thousands of symbols to display their contents.
+They need complex encoding standards such as `Unicode`_ and concepts like
+"character" no longer make sense. Instead, you have to deal with these terms:
+
+* `Code points`_: they are the atomic unit of information. A string is a series
+  of code points. Each code point is a number whose meaning is given by the
+  `Unicode`_ standard. For example, the English letter ``A`` is the ``U+0041``
+  code point and the Japanese *kana* ``の`` is the ``U+306E`` code point.
+* `Grapheme clusters`_: they are a sequence of one or more code points which are
+  displayed as a single graphical unit. For example, the Spanish letter ``ñ`` is
+  a grapheme cluster that contains two code points: ``U+006E`` = ``n`` (called
+  *"latin small letter N"*) + ``U+0303`` = ``◌̃`` (called *"combining tilde"*).
+* Bytes: they are the actual information stored for the string contents. Each
+  code point can require one or more bytes of storage depending on the standard
+  being used (UTF-8, UTF-16, etc.).
+
+The following image displays the bytes, code points and grapheme clusters for
+the same word written in English (``hello``) and Hindi (``नमस्ते``):
+
+.. image:: /_images/components/string/bytes-points-graphemes.png
+   :align: center
+
+Usage
+-----
+
+Create a new object of type :class:`Symfony\\Component\\String\\ByteString`,
+:class:`Symfony\\Component\\String\\CodePointString` or
+:class:`Symfony\\Component\\String\\UnicodeString`, pass the string contents as
+their arguments and then use the object-oriented API to work with those strings::
+
+    use Symfony\Component\String\UnicodeString;
+
+    $text = (new UnicodeString('This is a déjà-vu situation.'))
+        ->trimEnd('.')
+        ->replace('déjà-vu', 'jamais-vu')
+        ->append('!');
+    // $text = 'This is a jamais-vu situation!'
+
+    $content = new UnicodeString('नमस्ते दुनिया');
+    if ($content->lower()->startsWith('नमस्ते')) {
```

Comment:
`if ($content->ignoreCase()->startsWith('नमस्ते')) {` would be an even better example